### PR TITLE
Add environment variable to restrict certain keytypes

### PR DESCRIPTION
**Pull request by glasnt**
Per request in http://lists.lca2015.linux.org.au/pipermail/chat/2014-December/000118.html, I've added optional functionality to restrict weaker keys. Logic around finding the keytype follows. This may not be the best way to find out this information, but I couldn't find a saner alternative that didn't require importing possibly-bad keys into the keyring, then removing them again. 

---

The gpg command `--with-fingerprint` allows the parsing of an
ASCII-armored public key without the importing of said key into the
keyring. This can be used to check the strength of a key, and possibly
reject it from being submitted.

Sample fingerprint header: `pub  1024R/AABBCCDDEE 2011-11-11`
In this case, the type is parsed as `1024R`.

Environment variable format is any type of `^pub(\s*)(.+?)/` to
disclude. Example: `1024D,2048R` will reject 1024-DSA and 2048RDA keys.
In the above sample fingerprint header, this key would be rejected.

By default, all key types are allowed.

### CODE REVIEW

**Comment by frasertweedale** [comment]
Brilliant!  Will test this on the weekend.

**Comment by frasertweedale** [comment]
Actually I looked at it just now :)

Different versions of GnuPG report the key type differently, e.g. 2.1 says:

```
pub   rsa4096/4111E1E2 2011-11-18 [expires: 2016-11-16]
```

So, could you please use the `--with-colons` argument and grub the data out of there instead?
See http://git.gnupg.org/cgi-bin/gitweb.cgi?p=gnupg.git;a=blob;f=doc/DETAILS;h=ba2725fdbbd84dc5fd2b87c1e7e09a0ecbf5104b;hb=HEAD for details of the `--with-colons`
format and https://github.com/frasertweedale/gcaff/blob/v0.3/gcaff/gpg.py#L328-L332 for
an example of pulling out the algorithm and key size (or elliptic curve).

Thanks heaps for answering the call on this feature, Katie!

**Comment by glasnt** [comment]
Ah, knew there would have been a better way! I'll get on it.

**Comment by glasnt** [comment]
I've updated this to use the `--with-colon` logic. Should be a little bit more solid now.

File: pgpsubmitlib/pgp.py [hunk 1]
```diff
@@ -85,6 +109,41 @@ def import_keys(self, text):
...
+        return {"algorithm": algorithm, "length": length}
+
+    def valid_key(self,text):
+        """ Return true if key_type of text is not on the restricted list """
+        key = self.inspect_key(text)
+        restricted = self._environ.get("RESTRICTEDALGORITHMS","").split(",")
+        minimum_length = int(self._environ.get("MINIMUMLENGTH",0))
+        if key["algorithm"] in restricted:
+            return False, "Restricted Algorithm: %s" % PUB_ALGS[int(key["algorithm"])]
+        if key["length"] < minimum_length:
```
Comment (frasertweedale) [inline pgpsubmitlib/pgp.py:145]: I think this will fail for EC keys.  `length = pub[16]` from L129 is a misnomer; it gives the (string) curve name (elliptic curve keys do not have "bit-length" but rather an elliptic curve formula) which is not comparable with `int`.  I accept that I am responsible for this misnomer which appears in frasertweedale/gcaff :)

Aside from this, even for DSA/RSA keys I think that `key["length"]` remains a string at this point (but I haven't tried it so I could have missed something).

In summary: EC complicates key processing.

There are a couple of ways forward:
- Switch to a whitelist for algs, and a whitelist for lengths and curves.  Reasonable since there are only a few key lengths for DSA/RSA.
- Continue using bit-length threshold but map curve names to their equivalent RSA bit-length, based on https://tools.ietf.org/html/rfc6637#section-13.  This is probably the least work from where the patch is at now.

**Comment by glasnt** [comment]
I've adjusted the length check to map to the ECC Curve equiv length, as suggested. I don't think that having a whitelist is as maintainable going forward as having the option of a restricted list.

**Comment by frasertweedale** [comment]
Fair enough.  Ready for testing I'd say :)

...which I may not get to for a few days or so, Christmas time 'n' all...

**Comment by glasnt** [comment]
Fair enough. 

I have done a smidge of testing while iterating over this -- throwing a few different keys at it of different types, 1024-DSA, 4096-RSA, 256-ECC; using gpg and gpg2 executables -- though I'm sure you have more comprehensive methods.
